src/builder/build_traits.rs

```diff
@@ -61,11 +61,10 @@ pub trait Container {
     ///
     /// This function will return an error if there is an error in adding the
     /// [`OpType::Const`] node.
-    fn add_constant(&mut self, val: ConstValue) -> Result<ConstID, BuildError> {
-        let typ = val.const_type();
-        let const_n = self.add_child_op(ops::Const::new(val).map_err(BuildError::BadConstant)?)?;
+    fn add_constant(&mut self, constant: ops::Const) -> Result<ConstID, BuildError> {
```

Comment:
So my first thought was - "ah-ha! do we even need `add_constant`?" but now I see we kinda need it to control ConstID.

A possible future refactor, then. Imagine
```
trait NodeTemplate { // or BuildableNode, or something
   const op: OpTag;
   type op_ty: OpType; // probably only one of these
   ....whatever else we need....
}

impl NodeTemplate for Const {
  type op_ty = ops::Constant;
}

struct/?? NodeID<T> { // or maybe T is a "type", not sure
   fn get_node(&self) -> T {...}
}
```
and then `add_op<T: NodeTemplate>(n: T) -> NodeID<T>`

if you see what I mean
```